turbopack/crates/turbopack/src/global_module_ids.rs

```diff
@@ -0,0 +1,147 @@
+                let hash = hash_xxh3_hash64(&ident_str);
+                Ok((ident, (ident_str, hash)))
+            })
+            .try_join()
+            .await?
+            .into_iter()
+            .collect::<FxHashMap<_, _>>();
+
+        finalize_module_ids(&mut module_id_map);
+
+        Ok(GlobalModuleIdStrategy {
+            module_id_map: module_id_map
+                .into_iter()
+                .map(|(ident, (_, hash))| (ident, hash))
+                .collect(),
+        }
+        .cell())
+    }
+    .instrument(span)
+    .await
+}
+
+const JS_MAX_SAFE_INTEGER: u64 = (1u64 << 53) - 1;
+
+/// Shorten hashes and handle any collisions.
+fn finalize_module_ids(
+    merged_module_ids: &mut FxHashMap<ResolvedVc<AssetIdent>, (ReadRef<RcStr>, u64)>,
+) {
+    // 5% fill rate, as done in Webpack
+    // https://github.com/webpack/webpack/blob/27cf3e59f5f289dfc4d76b7a1df2edbc4e651589/lib/ids/IdHelpers.js#L366-L405
+    let optimal_range = merged_module_ids.len() * 20;
+    let digit_mask = std::cmp::min(
+        10u64.pow((optimal_range as f64).log10().ceil() as u32),
+        JS_MAX_SAFE_INTEGER,
+    );
+
+    // Run in multiple passes, to not depend on the order of the `merged_module_ids` (i.e. the order
+    // of imports). Hashes could still change if modules are added or removed.
+
+    // First pass: shorten hashes, potentially causing (more) collisions.
+    // used_ids contains all modules that will have to be rehashed in the next pass.
+    let mut used_ids =
+        FxHashMap::<u64, SmallVec<[(ResolvedVc<AssetIdent>, ReadRef<RcStr>); 1]>>::default();
+    for (ident, (ident_str, full_hash)) in merged_module_ids.iter_mut() {
+        let first_pass_hash = *full_hash % digit_mask;
+        match used_ids.entry(first_pass_hash) {
+            Entry::Vacant(e) => {
+                // First module using this hash, insert empty list. But the module can keep using
+                // this hash, no need to rehash later.
+                e.insert(SmallVec::default());
+            }
+            Entry::Occupied(mut e) => {
+                // A collision, insert module so that we can rehash it later.
+                e.get_mut().push((*ident, ident_str.clone()));
+            }
+        }
+
+        *full_hash = first_pass_hash;
+    }
+
+    // Filter conflicts
+    let mut conflicting_hashes = used_ids
+        .iter()
+        .filter(|(_, list)| !list.is_empty())
+        .map(|(hash, _)| *hash)
+        .collect::<Vec<_>>();
+    conflicting_hashes.sort();
+
+    // Second pass over the conflicts to resolve them
+    for hash in conflicting_hashes.into_iter() {
+        let list = used_ids.get_mut(&hash).unwrap();
+        // Take the vector but keep the (empty) entry, so that the "contains_key" check below
+        // doesn't reuse to original colliding hash.
+        let mut list = std::mem::take(list);
+        list.sort_by(|a, b| a.1.cmp(&b.1));
+
+        for (ident, _) in list.into_iter() {
```

Comment:
The first item can get the conflicting hash. It's already in used_ids.

```suggestion
        for (ident, _) in list.into_iter().skip(1) {
```